modules/Bio/EnsEMBL/Utils/ColourMap.pm

```diff
@@ -0,0 +1,304 @@
+    ## Optionally, add drawing code colours from Ensembl "stylesheet"
+    my %ensembl_colours = qw(
+      CONTRAST_BORDER   background0
+      CONTRAST_BG       background3
+
+      IMAGE_BG1         background1
+      IMAGE_BG2         background2
+
+      CONTIGBLUE1       contigblue1
+      CONTIGBLUE2       contigblue2
+
+      HIGHLIGHT1        highlight1
+      HIGHLIGHT2        highlight2
+    );
+
+    while (my($k,$v) = each %{$species_defs->ENSEMBL_STYLE||{}} ) {
+      my $k2 = $ensembl_colours{ $k };
+      next unless $k2;
+      $self->{$k2} = $v;
+    }
+  }
+
+  return $self;
+}
+
+sub is_defined {
+  my ($self, $name) = @_;
+  return exists $self->{$name};
+}
+
+sub hex_by_name {
+  my ($self, $name) = @_;
+  return '#' . ($self->{$name} || (
+    $name =~ /^#?([0-9a-fA-F]{6})$/ ? $1 : 
+    $name =~ /^(\d+),(\d+),(\d+)$/  ? sprintf '%02x%02x%02x', $1, $2, $3 : 'ff0000'
+  )); 
+}
+
+sub rgb_by_name {
+### Convert the provided colour to RGB
+### Note that despite the method name, it can accept a colour in hex format
+### or stringified RGB, as well as by Unix colour name
+### @param colour String - colour to be converted 
+### @param flag Boolean - ??
+### @return Array - RGB values
+  my ($self, $colour, $flag) = @_;
+  $colour = lc($colour);
+  my $hex;
+  
+  if ($colour =~ /(\d+,\d+,\d+)/) { ## RGB
+    return split(/,/, $1);
+  }
+  elsif ($colour =~ /^#?([0-9a-f]{6})$/) { ## Hex
+    $hex = $1;
+  }
+  else { ## Colour name - look up from list
+    $hex = $self->{$colour};
+  }
+
+  if ($hex) {
+    return $self->rgb_by_hex($hex);
+  }
+  else {
+    warn "Unknown colour name {$colour}" unless $errors->{$colour};
+    $errors->{$colour} = 1;
+    return $flag ? () : (0,0,0);
+  }
+
+}
+
+
+sub names {
+    my ($self) = @_;
+    return values %{$self};
+}
+
+sub hex_by_rgb {
+  my ($self, $triple_ref) = @_;
+  return sprintf("%02x%02x%02x", @{$triple_ref});
+}
+
+sub mix {
+  my( $self, $colour1, $colour2, $ratio ) = @_;
+  my @c1 = $self->rgb_by_name( $colour1, 1 );
+  my @c2 = $self->rgb_by_name( $colour2, 1 );
+  if( @c1 && @c2 ) {
+    return $self->hex_by_rgb([
+      $c1[0] + $ratio * ($c2[0]-$c1[0]),
+      $c1[1] + $ratio * ($c2[1]-$c1[1]),
+      $c1[2] + $ratio * ($c2[2]-$c1[2])
+    ]);
+  } elsif( @c1 ) {
+    return $colour1;
+  } elsif( @c2 ) {
+    return $colour2;
+  } else {
+    return 'red';
+  }
+}
+sub rgb_by_hex {
+    my ($self, $hex) = @_;
+
+    unless($hex){
+      warn "Cannot map hex colour in colourmap\n";
+      return(hex(0), hex(0), hex(0));
+    }
+    my ($hred, $hgreen, $hblue) = unpack("A2A2A2", $hex);
+    return (hex($hred), hex($hgreen), hex($hblue));
+}
+
+sub tint_by_hex {
+    my ($self, $hex, $rtone, $gtone, $btone) = @_;
+    return $self->hex_by_rgb([
+        $self->tint_by_rgb([ $self->rgb_by_hex($hex) ], $rtone, $gtone, $btone)
+    ]);
+}
+
+sub brightness {
+    my ($self, $name) = @_;
+    my ($r, $g, $b);
+    if ($name =~ /^#?([0-9a-f]{6})$/) { #Hex
```

Comment:
Haven't we got a helper function which tests this, and which handles both upper- and lowercase hexadigits?

Reply:
That's webteam's fault. Copy-pasted from the dawn of history most likely.